include/eve/module/core/regular/impl/simd/riscv/simd_cast.hpp

```diff
@@ -79,21 +82,63 @@ EVE_FORCEINLINE logical<wide<U, M>>
                            const O&,
                            logical<wide<T, N>>            x,
                            as<logical<wide<U, M>>> const                &tgt) noexcept
-requires(rvv_abi<abi_t<T, N>> || rvv_abi<abi_t<U, M>>)
+requires(rvv_abi<abi_t<T, N>> && rvv_abi<abi_t<U, M>>)
+{
+  using part_type_cast = detail::rvv_m1_wide<std::uint32_t>;
+  auto u_casted_in     = rvv_simd_cast(x, as<part_type_cast> {});
+  auto to_ret          = rvv_simd_cast(u_casted_in, tgt);
+  return to_ret;
+}
+
+template<typename N, typename M>
+EVE_FORCEINLINE wide<std::int8_t, M>
+                rvv_lmul_trunc(wide<std::int8_t, N> a, as<wide<std::int8_t, M>>) noexcept
+requires rvv_abi<abi_t<std::int8_t, N>>
+{
+  using out_wide_t        = wide<std::int8_t, M>;
+  constexpr auto out_lmul = rvv_lmul_v<std::int8_t, M>;
+  if constexpr( out_lmul == -8 ) return __riscv_vlmul_trunc_i8mf8(a);
+  else if constexpr( out_lmul == -4 ) return __riscv_vlmul_trunc_i8mf4(a);
+  else if constexpr( out_lmul == -2 ) return __riscv_vlmul_trunc_i8mf2(a);
+  else if constexpr( out_lmul == 1 ) return __riscv_vlmul_trunc_i8m1(a);
+  else if constexpr( out_lmul == 2 ) return __riscv_vlmul_trunc_i8m2(a);
+  else if constexpr( out_lmul == 4 ) return __riscv_vlmul_trunc_i8m4(a);
+}
+
+template<typename N, typename M>
+EVE_FORCEINLINE wide<std::int8_t, M>
+                rvv_lmul_extend(wide<std::int8_t, N> a, as<wide<std::int8_t, M>>) noexcept
+requires(rvv_abi<abi_t<std::int8_t, N>> && rvv_abi<abi_t<std::int8_t, M>>)
+{
+  using out_wide_t        = wide<std::int8_t, M>;
+  constexpr auto out_lmul = rvv_lmul_v<std::int8_t, M>;
+  if constexpr( out_lmul == -4 ) return __riscv_vlmul_ext_i8mf4(a);
+  else if constexpr( out_lmul == -2 ) return __riscv_vlmul_ext_i8mf2(a);
+  else if constexpr( out_lmul == 1 ) return __riscv_vlmul_ext_i8m1(a);
+  else if constexpr( out_lmul == 2 ) return __riscv_vlmul_ext_i8m2(a);
+  else if constexpr( out_lmul == 4 ) return __riscv_vlmul_ext_i8m4(a);
+  else if constexpr( out_lmul == 8 ) return __riscv_vlmul_ext_i8m8(a);
+}
+
+template<callable_options O, arithmetic_simd_value In, arithmetic_simd_value Out>
+EVE_FORCEINLINE Out
+simd_cast_(EVE_REQUIRES(rvv_), const O&, In x, as<Out> const& tgt) noexcept
+requires(rvv_abi<typename In::abi_type> && rvv_abi<typename Out::abi_type>)
 {
-  if constexpr( is_aggregated_v<abi_t<T, N>> || is_aggregated_v<abi_t<U, M>> )
+  constexpr auto size_cmp = sizeof(In) <=> sizeof(Out);
```

Comment:
just inline.

Reply:
You mean to use `sizeof(T) == sizeof(Out)` and `sizeof(T) < sizeof(Out)`?

Reply:
yeah

Reply:
Done